text/0095-type-aware-partial-evaluation.md

```diff
@@ -0,0 +1,471 @@
+
+In pseudocode, we describe typed partial requests as follows:
+
+```
+structure PartialRequest where
+  principal : PartialEntityUID
+  action : EntityUID                 -- Action is always known.
+  resource : PartialEntityUID
+  context : Option (Map Attr Value)  -- Unknown context is omitted.
+
+structure PartialEntityUID where
+  ty : EntityType                    -- Entity type is always known,
+  id : Option String                 -- but entity id may not be.
+```
+
+In JSON, we describe a partial request as an object with the optional `context` field and required `principal`, `action`, and `resource` fields. If present, the `context` value is specified according to the [Cedar JSON format](https://docs.cedarpolicy.com/auth/entities-syntax.html#context). If absent, it's treated as unknown and its type is inferred from the `action` and schema. The `principal` and `resource` fields are [entity objects](https://docs.cedarpolicy.com/auth/entities-syntax.html#uid) that can omit the `id` field to specify that the entity is unknown.
+
+For example, this partial request for our [basic schema](#basic-example) makes both the `context` and `resource` unknown:
+
+```
+{
+  "principal": { "type": "User", "id": "Alice" },
+  "action":    { "type": "Action", "id": "View" },
+  "resource":  { "type": "Document" }
+}
+```
+
+The key difference between this interface and CPE is that `context` must be either fully known or fully unknown. Partially specified `context` is not directly supported (though it can be expressed indirectly, as we'll see later). This design choice is consistent with classic partial evaluation, where input variables are either unknown or fully specified, and it simplifies implementation and formal modeling of TPE.
+
+#### Partial entities
+
+TPE specifies partial entities as a map from entity UIDs to partial entity data, consisting of unknown or fully specified attributes, ancestors, and tags. Unknown parts of the data are simply omitted, and their type is determined from the schema.
+
+In pseudocode, we describe partial entities as follows:
+
+```
+abbrev PartialEntities := Map EntityUID PartialEntityData
+
+structure PartialEntityData where
+  attrs : Option (Map Attr Value)    -- Attributes fully known or fully unknown.
+  ancestors : Option (Set EntityUID) -- Ancestors fully known or fully unknown.
+  tags : Option (Map String Value)   -- Tags are fully known or fully unknown.
+```
+
+In JSON, we describe partial entities as a list of JSON objects whose known parts are specified using the [Cedar JSON format for entities](https://docs.cedarpolicy.com/auth/entities-syntax.html#entities). Note that the ancestor hierarchy of an entity must be either unknown or fully specified.  So, the "parents" field of an entity cannot include an entity whose "parents" are unknown.
```

Comment:
If you have the energy, you might want to put the rewrite that we discussed on the PR comments in an appendix, so it's easy for people to see/understand/discover it from the proposal. 😄